### DISCUSSION TITLE
not init property

### DISCUSSION THREAD

**Discussion by Terry1511512** [dotnet/csharplang#4202] [Language Ideas]
I would like to see a "not init" property which forbids a property being used during initialisation.
Often a property is only usable after initialisation. If used in initialisation it then becomes dependent on initialisation order.

public class Idea 
{  
    public string Value { get; set; } 
    public string SomeValue { get; set; }  // Depends upon Value being initialised
}  

var Idea = new Idea  // Works
{
    Value = "XXXX",
   SomeValue = "YYYY"
};

var Idea = new Idea  // Won't Work
{
   SomeValue = "YYYY",   // Value not initialised
    Value = "XXXX"
};

A "not init" property that forbids the property being used in initialisation would stop this happening.

public class Idea 
{  
    public string Value { get; set; } 
    public string SomeValue { get; not init; }  // Depends upon Value being initialised
}

**Comment by theunrepentantgeek** [comment]
The usual response from the LDM when someone wants the language to prevent certain scenarios is to suggest writing an analyzer that implements your specific needs.

**Comment by Unknown6656** [comment]
Why not use a instance method as "setter" for that?
That way, all calls are guaranteed to be executed after the initialization performed by the constructor.
```csharp
public class Idea
{
     private string _some_value;

     public string Value { get; set; }
     public string SomeValue => _some_value; // getter always accessible


     public Idea() => _some_value = "Hello World!";

     public void SetSomeValue(string value)
     {
          // .... some logic
          _some_value = value;
     }
}
```

However, If you like the assignment syntax or like to use some operators, I'd advise you to use `ref` returns:
```csharp
public class Idea
{
     private string _some_value;

     public string Value { get; set; }

     public Idea() => _some_value = "Hello World!";

     public ref string SomeValue() => ref _some_value;
}
```
Which can be used thus:
```csharp
Idea i = new Idea { Value = "value" };

i.SomeValue() = "42";
string s = i.SomeValue();
```

**Reply by Terry1511512** [reply]
Of course I can always use methods but that's changing the usage away from a property syntax. I still want to use it as a normal property, just not in initialisation.

**Reply by Joe4evr** [reply]
Then writing an analyzer that ships in your library is basically your only option.
